common/src/main/java/com/scottlogic/deg/generator/fieldspecs/whitelist/FrequencyDistributedSet.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2019 Scott Logic Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.scottlogic.deg.generator.fieldspecs.whitelist;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class FrequencyDistributedSet<T> implements DistributedSet<T> {
+
+    private static final FrequencyDistributedSet<?> EMPTY = new FrequencyDistributedSet<>(Collections.emptySet());
+
+    private final Set<WeightedElement<T>> underlyingWeights;
+
+    private List<WeightedElement<T>> underlyingCumulativeWeights;
+
+    public FrequencyDistributedSet(final Set<WeightedElement<T>> underlyingWeights) {
+        if (underlyingWeights.isEmpty()) {
+            this.underlyingWeights = underlyingWeights;
+        } else {
+            if (underlyingWeights.contains(null)) {
+                throw new IllegalArgumentException("DistributedSet should not contain null elements");
+            }
+
+            double total = total(underlyingWeights);
+
+            this.underlyingWeights = underlyingWeights.stream()
+                .map(holder -> new WeightedElement<>(holder.element(), holder.weight() / total))
+                .collect(Collectors.toSet());
+        }
+    }
+
+    public static <T> double total(final Set<WeightedElement<T>> set) {
+        return set.stream()
+            .map(WeightedElement::weight)
+            .reduce(0.0D, Double::sum);
+    }
+
+    public static <T> FrequencyDistributedSet<T> uniform(final Set<T> underlyingSet) {
+        return new FrequencyDistributedSet<>(
+            underlyingSet.stream()
+                .map(e -> new WeightedElement<T>(e, 1.0D))
+                .collect(Collectors.toSet()));
+    }
+
+    private List<WeightedElement<T>> cumulative() {
+        if (underlyingCumulativeWeights == null) {
+            underlyingCumulativeWeights = generateCumulative(underlyingWeights);
```

Comment:
I figured that in the current implementation the class is being created much more often than the cumulative weights are being used.

I do agree that making the field is final is good to prevent possible reference changes - I was aiming to change this after the collection/stream interchanges elsewhere in the project were converted into singular streams, but I'm happy to change it now.